design/general-progress-monitoring.md

```diff
@@ -0,0 +1,469 @@
+written into the object store.  However, the actual data movement may be happening in the background
+after the backup has been marked "Completed".  The backup is not actually a stable backup until the
+data has been persisted properly.  In some cases (e.g. AWS) the backup cannot be restored from until
+the snapshots have been persisted.
+
+Once the snapshots have been taken, however, it is possible for additional backups or restores (as
+long as they don't use not-yet-completed backups) to be made without interference.  Waiting until
+all data has been moved before starting the next backup will slow the progress of the system without
+adding any actual benefit to the user.
+
+A new backup/restore phase, "WaitingForPluginOperations" will be introduced.  When a backup or
+restore has entered this phase, Velero is free to start another backup/restore.  The backup/restore
+will remain in the "WaitingForPluginOperations" phase until all BIA/RIA operations have completed
+(for example, for a volume snapshotter, until all data has been successfully moved to persistent
+storage).  The backup/restore will not fail once it reaches this phase.  If the backup is deleted
+(cancelled), the plug-ins will attempt to delete the snapshots and stop the data movement - this may
+not be possible with all storage systems.
+
+### State progression
+
+(Note that the diagrams have not yet been updated to reflect the terminology changes associated with
+the revised design.)
+
+![image](UploadFSM.png)
+### New
+When a backup/restore request is initially created, it is in the "New" phase.  
+
+The next state is either "InProgress" or "FailedValidation"
+
+### FailedValidation
+If the backup/restore request is incorrectly formed, it goes to the "FailedValidation" phase and
+terminates
+
+### InProgress
+When work on the backup/restore begins, it moves to the "InProgress" phase.  It remains in the
+"InProgress" phase until all pre/post execution hooks have been executed, all snapshots have been
+taken and the Kubernetes metadata and backup/restore info is safely written to the object store
+plug-in.
+
+In the current implementation, Restic backups will move data during the "InProgress" phase.  In the
+future, it may be possible to combine a snapshot with a Restic (or equivalent) backup which would
+allow for data movement to be handled in the "WaitingForPluginOperations" phase,
+
+The next phase is either "Completed", "WaitingForPluginOperations", "Failed" or "PartiallyFailed".
+Backups/restores which would have a final phase of "Completed" or "PartiallyFailed" may move to the
+"WaitingForPluginOperations" or "WaitingForPluginOperationsPartialFailure" state.  A backup/restore
+which will be marked "Failed" will go directly to the "Failed" phase.  Uploads may continue in the
+background for snapshots that were taken by a "Failed" backup/restore, but no progress will not be
+monitored or updated.  When a "Failed" backup is deleted, all snapshots will be deleted and at that
```

Comment:
So this is another part that I inherited from the original design. I suspect that it was because native snapshot uploads can't really be cancelled in progress. However, it might make sense to alter the design here to say that for Backup/RestoreItemActions in progress, the Cancel() API function will be called. Note that Cancel() isn't a guarantee, but if the plugin has a way to cancel the action, it will. What do others think about this change?

Reply:
I had similar concern before, but currently, backup only fails on validation stage before calling Plugins, so technically there should be no such case of failed backup with uploading or operation in background.
But we need to make sure this pattern will not be broken in future.

Reply:
Personally, I also vote on the change for this. Even though Velero doesn't reach the "failed" phase for now, from a common concept, the upload for a "failed" backup is meaningless. Suppose that we have 1TB of data to be uploaded, when a backup fails, Velero may start another backup shortly (maybe not now, but it is also a good way to handle a failed backup, otherwise, there will be recovery point miss), if we leave the upload for the failed backup and also start a new one, there will be double disk IO, double network IO, double cost (if any), and double or much longer completion time.
On the other hand, since Velero doesn't have the "failed" phase, we probably don't need to cover details of this "failed-cancel" in the design, just change the expression in this failed scenario.